File: Udemy/Data Structures/doublyLinkedList.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  push(val) {
    let newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
    this.length++;
    return this;
  }
  pop() {
    if (!this.head) return undefined;
    let poppedNode = this.tail;
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = poppedNode.prev;
      this.tail.next = null;
      poppedNode.prev = null;
    }
    this.length--;
    return poppedNode;
  }
  shift() {
    if (!this.head) return undefined;
    let poppedNode = this.head;
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = poppedNode.next;
      this.head.prev = null;
      poppedNode.next = null;
    }
    this.length--;
    return poppedNode;
  }
  unshift(val) {
    let newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.head.prev = newNode;
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
    return this;
  }
  get(index) {
    if (index < 0 || index >= this.length) return undefined;
    let countIndex, currentNode;
    if (index <= this.length / 2) {
      console.log("WORKING FROM START");
      countIndex = 0;
      currentNode = this.head;
      while (index !== countIndex) {
        countIndex++;
        currentNode = currentNode.next;
      }
    } else {
      console.log("WORKING FROM END");
      countIndex = this.length - 1;
      currentNode = this.tail;
      while (index !== countIndex) {
        countIndex--;
        currentNode = currentNode.prev;
      }
    }
    return currentNode;
  }
  set(index, val) {
    let validNode = this.get(index);
    if (validNode) {
      validNode.val = val;
      return true;
    }
    return false;
  }
  insert(index, val) {
    if (index < 0 || index > this.length) return false;
    if (index === 0) return !!this.unshift(val);
    if (index === this.length) return this.push(val);

    let newNode = new Node(val);
    let foundNode = this.get(index - 1);

    foundNode.next.prev = newNode;
    newNode.next = foundNode.next;

    foundNode.next = newNode;
    newNode.prev = foundNode;

    this.length++;
    return true;
  }
  remove(index) {
    if (index < 0 || index >= this.length) return undefined;
    if (index === 0) return this.shift();
    if (index === this.length - 1) return this.pop();
    let foundNode = this.get(index);
    let beforeNode = foundNode.prev;
    let afterNode = foundNode.next;

    beforeNode.next = foundNode.next;
    afterNode.prev = foundNode.prev;
    // foundNode.prev.next = foundNode.next;
    // foundNode.next.prev = foundNode.prev;

    foundNode.next = null;
    foundNode.prev = null;
    this.length--;
    return foundNode;
  }
}
let newDoubly = new DoublyLinkedList();
newDoubly.push(3);

newDoubly.push(4);

newDoubly.push(5);
console.log(newDoubly);
console.log("==============");
console.log(newDoubly.remove(1));
console.log("==============");
console.log(newDoubly);
console.log(newDoubly.remove(1));
console.log("==============");
console.log(newDoubly);

File: Algoexprt/doubly_linked_list.py
from abc import abstractmethod


class NoSuchElementException(Exception):
    pass


class WrongIndex(Exception):
    pass


class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class DoublyLinkedListInterface:
    @abstractmethod
    def add_first(self, node: Node) -> None:
        pass

    @abstractmethod
    def add_last(self, node: Node) -> None:
        pass

    @abstractmethod
    def remove_first(self) -> Node:
        pass

    @abstractmethod
    def remove_last(self) -> Node:
        pass

    @abstractmethod
    def search(self, index) -> Node:
        pass

    @abstractmethod
    def insert(self, index, node) -> None:
        pass

    @abstractmethod
    def remove(self, index) -> Node:
        pass


class DoublyLinkedList(DoublyLinkedListInterface):
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def add_first(self, node: Node) -> None:
        if self.size == 0:
            self.head = node
            self.tail = node
        else:
            self.head.prev = node
            node.next = self.head
            self.head = node
        self.size += 1
        return

    def add_last(self, node: Node) -> None:
        if self.size == 0:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            node.prev = self.tail
            self.tail = node
        self.size += 1
        return

    def remove_first(self) -> Node:
        if self.size == 0:
            raise NoSuchElementException()
        elif self.size == 1:
            node = self.head
            self.head = None
            self.tail = None
        else:
            node = self.head
            self.head = node.next
            node.next.prev = None
            node.next = None
        self.size -= 1
        return node

    def remove_last(self) -> Node:
        if self.size == 0:
            raise NoSuchElementException()
        elif self.size == 1:
            node = self.tail
            self.head = None
            self.tail = None
        else:
            node = self.tail
            self.tail = node.prev
            node.prev.next = None
            node.prev = None
        self.size -= 1
        return node

    def search(self, index) -> Node:
        if index < 0 or index > self.size - 1:
            raise WrongIndex()
        elif index < ((self.size - 1) / 2):
            foo_bar = 0
            node = self.head
            while foo_bar != index:
                foo_bar += 1
                node = node.next
            return node
        else:
            foo_bar = self.size - 1
            node = self.tail
            while foo_bar != index:
                foo_bar -= 1
                node = node.prev
            return node

    def insert(self, index, node) -> None:
        if index < 0 or index > self.size:
            raise WrongIndex()
        elif index == 0:
            self.add_first(node)
        elif index == self.size:
            self.add_last(node)
        else:
            found_node = self.search(index)
            prev_node = found_node.prev
            prev_node.next = node
            found_node.prev = node
            node.prev = prev_node
            node.next = found_node
            self.size += 1
        return

    def remove(self, index) -> Node:
        if index < 0 or index > self.size - 1:
            raise WrongIndex()
        elif index == 0:
            return self.remove_first()
        elif index == self.size - 1:
            return self.remove_last()
        else:
            node = self.search(index)
            next_node = node.next
            prev_node = node.prev
            next_node.prev = prev_node
            prev_node.next = next_node
            node.next = None
            node.prev = None
            self.size -= 1
            return node

    def __str__(self):
        node = self.head
        result = "None "
        while node is not None:
            result += "<- "
            result += str(node.value)
            result += " -> "
            node = node.next
        result += "None"
        return result


node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)
node5 = Node(5)

linked_list = DoublyLinkedList()

linked_list.add_first(node1)
linked_list.add_last(node5)
linked_list.insert(1, node2)
linked_list.insert(2, node3)
linked_list.insert(3, node4)

linked_list.remove(3)
linked_list.remove(3)

print(linked_list)
